src/mpi/coll/op/oputil.c

```diff
@@ -74,3 +74,128 @@ const char *MPIR_Op_builtin_get_shortname(MPI_Op op)
     }
     return "";
 }
+
+int MPIR_Reduce_equal(const void *sendbuf, MPI_Aint count, MPI_Datatype datatype, int root,
+        MPIR_Comm * comm_ptr, MPIR_Errflag_t * errflag)
+{
+	/*
+		IMPORTANT:
+		the sendbuf has to be a stuct in the following format
+		struct stuct_name{
+			datatype data[count]; 	//this count is also used as the input parameter count
+									//datatype has to be char, int, or long
+									//for any other types, if you beleive memcmp() gives correct comparision, 
+									//you may call this func at own risk
+			int is_equal;
+		}
+	 */
+    MPI_Op op = MPIX_EQUAL;
+    int mpi_errno = MPI_SUCCESS;
+    MPI_Aint type_size, is_equal_pos, lb, extent;
+    void *recvbuf;
+
+    mpi_errno = MPIR_Type_get_extent_impl(datatype, &lb, &extent);
+    if(mpi_errno != MPI_SUCCESS || lb != 0)//Not able to handle structs not starting at 0
+    	return mpi_errno;
+
+    type_size = extent - lb;
+
+
+    //Dealing simple c stuct alignment and padding.
+    if ((type_size % sizeof(int)) == 0) 
+    {
+    	is_equal_pos = type_size * count;
+	} 
+	else 
+	{
+    	is_equal_pos = type_size * count + (sizeof(int) - type_size % sizeof(int));
+	}
+
+    int s_count = 2;
+    int s_blocklengths[2] = {count, 1};
+    MPI_Datatype s_types[2] = {datatype, MPI_INT};
+    MPI_Datatype derived_type;  
+
+    MPI_Aint s_displacements[2] = {0, is_equal_pos};
+
+    /* set up type */
+    mpi_errno = MPI_Type_create_struct(s_count, s_blocklengths, s_displacements, s_types, &derived_type);
+    MPI_Type_commit(&derived_type);
+
+    *(int*)(sendbuf + is_equal_pos) = 1;
+
+    mpi_errno = MPIR_Type_get_extent_impl(derived_type, &lb, &extent);
+
+    recvbuf = MPL_malloc(extent-lb, MPL_MEM_OTHER);
+
+    mpi_errno = MPIR_Reduce_impl(sendbuf, recvbuf, 1, derived_type, MPIX_EQUAL, root, comm_ptr, errflag);
+
+    *(int*)(sendbuf + is_equal_pos) = *(int*)(recvbuf + is_equal_pos);
+
+    MPL_free(recvbuf);
+    MPI_Type_free(&derived_type);
+    return mpi_errno;
+}
+
+
+int MPIR_Allreduce_equal(const void *sendbuf, MPI_Aint count, MPI_Datatype datatype,
+        MPIR_Comm * comm_ptr, MPIR_Errflag_t * errflag)
+{
+	/*
+		IMPORTANT:
+		the sendbuf has to be a stuct in the following format
+		struct stuct_name{
+			datatype data[count]; 	//this count is also used as the input parameter count
+									//datatype has to be char, int, or long
+									//for any other types, if you beleive memcmp() gives correct comparision, 
+									//you may call this func at own risk
+			int is_equal;
```

Comment:
This is different from what I thought we agreed upon. I thought we were implementing --
```
int MPIR_Allreduce_equal(const void *buf, MPI_Aint count, MPI_Datatype datatype, int *is_equal, MPIR_Comm *comm_ptr)
```
Where there is no restriction on the datatype. Then as an implementation detail, we may start with only support a basic integer type for datatype and return error for all other datatypes.

Reply:
Do you want to have "int  *is_equal" as a parameter? If the *buf is a stuct, it should have a member is_equal.

Reply:
The whole point of having this function is so the user don't have to go through the datatype creations. The hidden requirement of "should" is a very bad interface.

Reply:
Required input parameters changed to `int MPIR_Allreduce_equal(const void *buf, MPI_Aint count, MPI_Datatype datatype, int *is_equal, MPIR_Comm *comm_ptr, MPIR_Errflag_t *errflag) `